### ISSUE TITLE
`Error: The configmap "aws-auth" does not exist` when deploying an EKS cluster with `manage_aws_auth_configmap = true`

### ISSUE THREAD

**Issue by ezzatron** [terraform-aws-modules/terraform-aws-eks#2009]
## Description

When deploying an EKS cluster using `manage_aws_auth_configmap = true`, the deploy fails with the error:

    Error: The configmap "aws-auth" does not exist

- [x] ✋ I have searched the open/closed issues and my issue is not listed.

## Versions

- Module version [Required]: `~> 18.20.1`

- Terraform version: Terraform Cloud
- Provider version(s):
  - `aws v4.9.0`
  - `cloudinit v2.2.0`
  - `kubernetes v2.10.0`
  - `tls v3.3.0`

## Reproduction Code [Required]

```hcl
data "aws_availability_zones" "available" {}

module "vpc_example" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.14.0"

  name = "example"
  cidr = "10.10.0.0/16"

  azs                = slice(data.aws_availability_zones.available.names, 0, 3)
  enable_nat_gateway = true

  private_subnets  = ["10.10.1.0/24", "10.10.2.0/24", "10.10.3.0/24"]
  public_subnets   = ["10.10.11.0/24", "10.10.12.0/24", "10.10.13.0/24"]
  database_subnets = ["10.10.21.0/24", "10.10.22.0/24", "10.10.23.0/24"]

  public_subnet_tags = {
    "kubernetes.io/role/elb" = 1
  }

  private_subnet_tags = {
    "kubernetes.io/role/internal-elb" = 1
  }
}

data "aws_iam_roles" "sso_breakglass" {
  name_regex  = "AWSReservedSSO_BreakGlass_.*"
  path_prefix = "/aws-reserved/sso.amazonaws.com/"
}
data "aws_iam_roles" "sso_readall" {
  name_regex  = "AWSReservedSSO_ReadAll_.*"
  path_prefix = "/aws-reserved/sso.amazonaws.com/"
}

module "eks_main" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 18.20.1"

  cluster_name                    = "main"
  cluster_enabled_log_types       = ["api", "audit", "authenticator", "controllerManager", "scheduler"]
  cluster_endpoint_private_access = true

  vpc_id     = module.vpc_example.vpc_id
  subnet_ids = module.vpc_example.private_subnets

  eks_managed_node_groups = {
    spot = {
      create_launch_template = false
      launch_template_name   = ""

      capacity_type = "SPOT"
      instance_types = [
        "m4.large",
        "m5.large",
        "m5a.large",
        "m6i.large",
        "t2.large",
        "t3.large",
        "t3a.large",
      ],
    }
  }

  manage_aws_auth_configmap = true

  aws_auth_roles = [
    {
      rolearn  = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/${one(data.aws_iam_roles.sso_breakglass.names)}"
      username = "sso-breakglass:{{SessionName}}"
      groups   = ["sso-breakglass"]
    },
    {
      rolearn  = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/${one(data.aws_iam_roles.sso_readall.names)}"
      username = "sso-readall:{{SessionName}}"
      groups   = ["sso-readall"]
    },
  ]
}
```

- Attempt to deploy an EKS cluster with EKS managed node groups and `manage_aws_auth_configmap = true`
- See the above error

In our case, we're using Terraform Cloud, but I'm unsure if that actually affects anything here. We were simply trying to create a new EKS cluster, and noticed this new setting. In the past we've had to use complex hacks to manage the `aws-auth` ConfigMap, so this seemed like a better approach, but it doesn't seem to work.

It's worth noting that running another apply doesn't fix the issue, so I don't think it's a timing issue.

## Expected behavior

The cluster is created without error, and the `aws-auth` ConfigMap contains the expected content.

## Actual behavior

The above error.

## Additional context

<img width="1398" alt="Screen Shot 2022-04-12 at 09 53 54" src="https://user-images.githubusercontent.com/100152/162851549-9cdf4308-2697-44bc-bb75-3537a6cb88bc.png">

**Comment by bryantbiggs** [comment]
thats odd that its stating that the `aws-auth` configmap doesn't exist when you are using an EKS managed node group - EKS managed node groups automatically create/update the configmap with the role used by the node group (same with Fargate profiles). However, self managed node groups do not create the configmap so we do have a variable to handle this

https://github.com/terraform-aws-modules/terraform-aws-eks/blob/69a815c7dfe3c44feddfaa129c36953656d5f123/examples/self_managed_node_group/main.tf#L61-L63

```hcl
create_aws_auth_configmap = true
manage_aws_auth_configmap = true
```

**Comment by ezzatron** [comment]
Yeah, I assumed based on the docs that we _shouldn't_ be setting `create_aws_auth_configmap = true` because we're only using managed node groups. Should enable that setting as well?

**Comment by bryantbiggs** [comment]
yes, its safe to enable this.

question: if the configmap doesn't exist, are your nodes connected to the control plane?

**Comment by ezzatron** [comment]
As far as I can tell, the ConfigMap genuinely doesn't exist. In the console, the node group shows up under the cluster, which (I think) means that the nodes are connected to the control plane:

<img width="1419" alt="Screen Shot 2022-04-12 at 10 35 47" src="https://user-images.githubusercontent.com/100152/162855215-61028ef2-51d7-49d0-82c4-e1d69ff95aa7.png">

But since the config map doesn't exist, I can't see any of the resources inside the cluster of course:

<img width="1419" alt="Screen Shot 2022-04-12 at 10 35 06" src="https://user-images.githubusercontent.com/100152/162855149-9fd5036f-0ebe-49f8-b774-15c81314dac1.png">

**Comment by bryantbiggs** [comment]
hmm, somethings off. if the configmap doesn't exist then the nodes won't register because they lack authorization to do so

**Comment by bryantbiggs** [comment]
let me try your repro and see

**Comment by ezzatron** [comment]
Thanks for taking a look, let me know if I can help out with any other info. In the meantime I'm going to try enabling `create_aws_auth_configmap = true`, and also seeing if the spot instance config has something to do with it (that's another thing we changed at the same time from a working cluster config).

**Comment by ezzatron** [comment]
FWIW, adding `create_aws_auth_configmap = true` did change the error we get, but it didn't help us understand what's going on:

    Error: Post "http://localhost/api/v1/namespaces/kube-system/configmaps": dial tcp 127.0.0.1:80: connect: connection refused

<img width="1381" alt="Screen Shot 2022-04-12 at 10 48 31" src="https://user-images.githubusercontent.com/100152/162856426-5a6b17bd-ecf2-40ca-9b2c-06e947895479.png">

**Comment by bryantbiggs** [comment]
hmm, it is there but its not recognizing or patching it. I'll have to file an issue with the upstream Kubernetes provider in the morning to have them take a look

**Comment by ezzatron** [comment]
No worries, thanks for your help 🙏

**Comment by lrstanley** [comment]
Actually, nevermind. It succeeded, however it didn't actually apply it.

**Comment by MadsRC** [comment]
I had the same issue, both the `config map does not exist` when managing the config map and the `connection refused 127.0.0.1` when attempting to create it.

I'm using managed node groups as well.

The way I solved it was to add a kubernetes provider. This here should be enough:
```HCL
/*
The following 2 data resources are used get around the fact that we have to wait
for the EKS cluster to be initialised before we can attempt to authenticate.
*/

data "aws_eks_cluster" "default" {
  name = module.eks.cluster_id
}

data "aws_eks_cluster_auth" "default" {
  name = module.eks.cluster_id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.default.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.default.token
}
```

It's also a great way to authenticate to the EKS cluster instead of the example in the repo that forces the use of `awscli`

**Comment by bryantbiggs** [comment]
> I had the same issue, both the `config map does not exist` when managing the config map and the `connection refused 127.0.0.1` when attempting to create it.
> 
> I'm using managed node groups as well.
> 
> The way I solved it was to add a kubernetes provider. This here should be enough:
> 
> ```terraform
> /*
> The following 2 data resources are used get around the fact that we have to wait
> for the EKS cluster to be initialised before we can attempt to authenticate.
> */
> 
> data "aws_eks_cluster" "default" {
>   name = module.eks.cluster_id
> }
> 
> data "aws_eks_cluster_auth" "default" {
>   name = module.eks.cluster_id
> }
> 
> provider "kubernetes" {
>   host                   = data.aws_eks_cluster.default.endpoint
>   cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
>   token                  = data.aws_eks_cluster_auth.default.token
> }
> ```
> 
> It's also a great way to authenticate to the EKS cluster instead of the example in the repo that forces the use of `awscli`

oy, good spot on not providing the provider and creds. I'll file a ticket for better error reporting on that upstream. 

Regarding the token vs exec - exec is what is recommended by the provider itself https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#exec-plugins

**Comment by bryantbiggs** [comment]
yep, was able to confirm that was the issue and this now works as expected @ezzatron - we just forgot to put the provider auth 

```hcl
provider "aws" {
  region = "us-east-1"
}

provider "kubernetes" {
  host                   = module.eks_main.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks_main.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks_main.cluster_id]
  }
}

data "aws_availability_zones" "available" {}

data "aws_caller_identity" "current" {}

module "vpc_example" {
  source  = "terraform-aws-modules/vpc/aws"
  version = "~> 3.14.0"

  name = "example"
  cidr = "10.10.0.0/16"

  azs                = slice(data.aws_availability_zones.available.names, 0, 3)
  enable_nat_gateway = true

  private_subnets  = ["10.10.1.0/24", "10.10.2.0/24", "10.10.3.0/24"]
  public_subnets   = ["10.10.11.0/24", "10.10.12.0/24", "10.10.13.0/24"]
  database_subnets = ["10.10.21.0/24", "10.10.22.0/24", "10.10.23.0/24"]

  public_subnet_tags = {
    "kubernetes.io/role/elb" = 1
  }

  private_subnet_tags = {
    "kubernetes.io/role/internal-elb" = 1
  }
}

module "eks_main" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 18.20.1"

  cluster_name                    = "main"
  cluster_enabled_log_types       = ["api", "audit", "authenticator", "controllerManager", "scheduler"]
  cluster_endpoint_private_access = true

  vpc_id     = module.vpc_example.vpc_id
  subnet_ids = module.vpc_example.private_subnets

  eks_managed_node_groups = {
    spot = {
      create_launch_template = false
      launch_template_name   = ""

      capacity_type = "SPOT"
      instance_types = [
        "m4.large",
        "m5.large",
        "m5a.large",
        "m6i.large",
        "t2.large",
        "t3.large",
        "t3a.large",
      ],
    }
  }

  manage_aws_auth_configmap = true

  aws_auth_roles = [
    {
      rolearn  = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/foo"
      username = "sso-breakglass:{{SessionName}}"
      groups   = ["sso-breakglass"]
    },
    {
      rolearn  = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/bar"
      username = "sso-readall:{{SessionName}}"
      groups   = ["sso-readall"]
    },
  ]
}
```

**Comment by bryantbiggs** [comment]
@ezzatron are we able to close this with the solution posted above?

**Comment by bryantbiggs** [comment]
@narenaryan just be mindful that using that route the token can expire. The provider recommends the exec route if you can (requires awscli to be available where Terraform is execute) https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#exec-plugins

**Comment by magnusseptim** [comment]
It seems to still be some issue.

Tried both @bryantbiggs and @narenaryan propostions, and yes, it _sometimes_ works as intended.

Unfortunately from time to time I get :

` Error: Get "https://***.***.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp <ip address here>:443: i/o timeout
`

on terraform plan stage.

Was unable to find yet what was the underlying reason.

**Comment by bryantbiggs** [comment]
@magnusseptim please feel free to investigate, add a 👍🏽 , or post a new issue upstream as there are well known issues with the Kubernetes provider https://github.com/hashicorp/terraform-provider-kubernetes/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+localhost

**Comment by ncjones** [comment]
When configuring the kubernetes provider using exec and no region, as per the advice above, I got the following error:

```
module.eks.kubernetes_config_map.aws_auth[0]: Creating...
│ Error: Post "https://9266dd6a08gr7.us-west-2.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps": getting credentials: exec: executable aws failed with exit code 255
│ 
│   with module.eks.kubernetes_config_map.aws_auth[0],
│   on .terraform/modules/eks/main.tf line 414, in resource "kubernetes_config_map" "aws_auth":
│  414: resource "kubernetes_config_map" "aws_auth" {
│ 
Error: Process completed with exit code 1.
```
I think it's because it assumes the AWS CLI is configured with a region which is not true in my environment. If the region is not set then the AWS CLI will attempt to contact the instance metadata service (IMDS) to detect the region. The IMDS call also fails in my CI/CD environment.

Using the `aws_eks_cluster_auth` data source solves this issue.

```hcl
data "aws_eks_cluster_auth" "default" {
  name = "my-eks-cluster-name"
}

provider "kubernetes" {
  host = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token = data.aws_eks_cluster_auth.default.token
}
```

**Comment by bryantbiggs** [comment]
The exec command is sent to the awscli so you can set the region:

```hcl
provider "kubernetes" {
  host                   = module.eks_main.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks_main.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks_main.cluster_id, "--region", "us-east-1"]
  }
}
```

**Comment by Hokwang** [comment]
in my case, I have several profiles, so I need to add `"--profile"` option.

**Comment by magnusseptim** [comment]
So, eventually I was able to handle my issue.

Result :

In my case, a combination of  `cluster_endpoint_private_access = true`, `-parallelism=1` (in a form of env variable on my devops agent) and `
coredns = {
      resolve_conflicts = "OVERWRITE"
    }
` (but that one only cause of planned usage of fargate) worked.

How I was able to figure that out :

Overall, I think  `-parallelism=1` was crucial, as when I tried to dig-out what was the issue, I tried to 'reproduce' aws-auth applying procedure with older version of module (18.19.0) and kubernetes provider (1.13 / 1.19) on my own.

I did something similar to : https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1901#issuecomment-1060428753 and https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1744#issuecomment-1046810281 which ended with localhost-kind of issues, like in https://github.com/hashicorp/terraform-provider-kubernetes/issues/1028

Which lead me to `-parallelism=1`.

During tests also noticed that if both `cluster_endpoint_private_access=true` and `cluster_endpoint_public_access=false` are set, updating of cluster is 'unstable' (which I allow myself to name like that cause of lack of better word).

Setting those flags like that ends with random issues happening for both `token` and `exe` kubernetes provider approach, mostly failing as a 30+ sec. timeouts from config map endpoint `api/v1/namespaces/kube-system/configmaps`

Not sure about underlying issue, but I considered it worth mentioning here

| Terraform | 1.1.8 |
| EKS Module | 18.20.2 |
| Kubernetes Provider | 2.10.0 |

**Comment by ezzatron** [comment]
> @ezzatron are we able to close this with the solution posted above?

Thanks for the help - I'm pretty sure I wasn't using that type of auth so it seems like it would have fixed the original issue. I ended up using https://github.com/aws-ia/terraform-aws-eks-blueprints in the meantime, which means I'm no longer manually adding to `aws-auth`.

I'll close the issue in any case 👍

**Comment by guitmz** [comment]
this is still happening in the latest version, not sure why the issue was closed without a proper solution. It was not happening with older module versions

**Comment by kaykhancheckpoint** [comment]
I ran into this issue aswell today and it was because i had multiple profiles on my local machine, setting the profile worked for me .

```
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--profile", locals.aws_profile]
  }
}
```

**Comment by guitmz** [comment]
I'm still getting an error

```
│ Error: The configmap "aws-auth" does not exist
│
│   with kubernetes_config_map_v1_data.aws_auth[0],
│   on main.tf line 431, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  431: resource "kubernetes_config_map_v1_data" "aws_auth" {
│
```

I have

```hcl
manage_aws_auth_configmap = true
  aws_auth_roles = [
    {
      rolearn  = include.root.inputs.aws_sso_admin_role_arn
      username = "AdministratorSSORole:{{SessionName}}"
      groups   = ["system:masters"]
    },
  ]
```

and

```hcl
provider "kubernetes" {
  host                   = aws_eks_cluster.this[0].endpoint
  cluster_ca_certificate = base64decode(aws_eks_cluster.this[0].certificate_authority[0].data)
  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    args        = ["eks", "get-token", "--cluster-name", "infra-tools", "--profile", "my-profile-here"]
    command     = "aws"
  }
}
```

I also tried with `create_aws_auth_configmap  = true` but that makes it worse, I get the expected permission denied error.

**Comment by tculp** [comment]
@bryantbiggs 

So the terraform docs (https://www.terraform.io/language/providers/configuration) says that

```
You can use [expressions](https://www.terraform.io/language/expressions) in the values of these configuration arguments, but can only reference values that are known before the configuration is applied. This means you can safely reference input variables, but not attributes exported by resources (with an exception for resource arguments that are specified directly in the configuration).
```

However, the workaround with configuring the kubernetes provider from the output of the module is counter to this. Is the terraform docs just being overly paranoid, or can there sometimes be issues with the workaround?

**Comment by FernandoMiguel** [comment]
> @bryantbiggs 
> 
> So the terraform docs (https://www.terraform.io/language/providers/configuration) says that
> 
> ```
> You can use [expressions](https://www.terraform.io/language/expressions) in the 
> values of these configuration arguments, but can only reference values that are 
> known before the configuration is applied. This means you can safely reference 
> input variables, but not attributes exported by resources (with an exception for 
> resource arguments that are specified directly in the configuration).
> ```
> 
> However, the workaround with configuring the kubernetes provider from the output of the module is counter to this. Is the terraform docs just being overly paranoid, or can there sometimes be issues with the workaround?
> 
> If there are potential issues with this method, then I think that the `aws_auth_configmap_yaml` output should be un-deprecated

My experience with this shows that the kube provider does not know the cluster endpoint or valid token.
It's weird cause I was able to use those details in a local template.
I guess the provider is initialised early in the graph and lacks the correct values.

**Comment by flowinh2o** [comment]
> For any Terragrunt users, this block solves it.
> 
> ```
> generate "provider" {
>   path      = "provider.tf"
>   if_exists = "overwrite"
>   contents = <<EOF
> provider "kubernetes" {
>   host                   = aws_eks_cluster.this[0].endpoint
>   cluster_ca_certificate = base64decode(aws_eks_cluster.this[0].certificate_authority[0].data)
>   exec {
>     api_version = "client.authentication.k8s.io/v1alpha1"
>     args        = ["eks", "get-token", "--cluster-name", var.cluster_name]
>     command     = "aws"
>   }
> }
> EOF
> }
> ```

Thank you.  This worked for me as well.

**Comment by dm-sumup** [comment]
Looks like this issue is closed, but none of the fixes provided here work.

Here is my tf file:

```
data "aws_eks_cluster" "default" {
  name = module.eks.cluster_id
}

data "aws_eks_cluster_auth" "default" {
  name = module.eks.cluster_id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.default.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    args        = ["eks", "get-token", "--cluster-name", var.cluster_name, "--profile", var.customer-var.environment]
    command     = "aws"
  }
  # token                  = data.aws_eks_cluster_auth.default.token
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 18.0"

  cluster_name    = var.cluster_name
  cluster_version = var.cluster_version
  
  cluster_endpoint_private_access = true
  cluster_endpoint_public_access  = false

  cluster_enabled_log_types = ["api", "audit", "authenticator", "controllerManager", "scheduler"]

  cluster_addons = {
    coredns = {
      resolve_conflicts = "OVERWRITE"
    }
    kube-proxy = {}
    vpc-cni = {
      resolve_conflicts = "OVERWRITE"
    }
  }
  
  vpc_id  = var.vpc_id
  subnet_ids = var.subnet_ids

  cluster_encryption_config = [{
    provider_key_arn = var.kms_key_id
    resources        = ["secrets"]
  }]

  # EKS Managed Node Group(s)
  eks_managed_node_group_defaults = {
    disk_size              = 50
    instance_types         = ["c5.large"]
  }

  eks_managed_node_groups = {
    "${var.ng1_name}" = {
      min_size     = var.ng1_min_size
      max_size     = var.ng1_max_size
      desired_size = var.ng1_desired_size

      instance_types = var.ng1_instance_types
      capacity_type  = "ON_DEMAND"
      
      update_config = {
        max_unavailable_percentage = 50
      }

      tags = var.tags
    }
  }

  node_security_group_additional_rules = var.ng1_additional_sg_rules

  # aws-auth configmap
  manage_aws_auth_configmap = true

  tags = var.tags
}
```

I still get 
```
│ Error: The configmap "aws-auth" does not exist
│ 
│   with module.eks-cluster.module.eks.kubernetes_config_map_v1_data.aws_auth[0],
│   on .terraform/modules/eks-cluster.eks/main.tf line 431, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  431: resource "kubernetes_config_map_v1_data" "aws_auth" {
```

Any ideas?

Using:
eks modue `~> 18.0`
terraform `>= 0.13.1`
aws `>= 3.35.0, < 4.0`
kubernetes `>= 2.10.0, < 2.12`

**Comment by dempti** [comment]
@bryantbiggs I am still faicing the same issue while setting `manage_aws_auth_configmap=true` it fails for managed node group, but it set `false` it passes

```
aws.com/id/B91CD86D1023F225A6AFAD3DD6515EBF]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # kubernetes_config_map_v1_data.aws_auth[0] will be created
  + resource "kubernetes_config_map_v1_data" "aws_auth" {
      + data  = {
          + "mapAccounts" = jsonencode([])
          + "mapRoles"    = <<-EOT
                - "groups":
                  - "system:bootstrappers"
                  - "system:nodes"
                  "rolearn": "arn:aws:iam::{{aws_id}}}:role/eks-managed-node-group"
                  "username": "system:node:{{EC2PrivateDNSName}}"
                - "groups":
                  - "system:masters"
                  "rolearn": "admin_role"
                  "username": "admin"
            EOT
          + "mapUsers"    = jsonencode([])
        }
      + force = true
      + id    = (known after apply)

      + metadata {
          + name      = "aws-auth"
          + namespace = "kube-system"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
kubernetes_config_map_v1_data.aws_auth[0]: Creating...
╷
│ Error: The configmap "aws-auth" does not exist
│
│   with kubernetes_config_map_v1_data.aws_auth[0],
│   on main.tf line 431, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  431: resource "kubernetes_config_map_v1_data" "aws_auth" {
│
╵
Releasing state lock. This may take a few moments...
ERRO[0079] Module /home/redemption/WORK/Learn/LEARN/pyrite/computer-vision/modules/eks has finished with an error: 1 error occurred:
	* exit status 1
  prefix=[/home/redemption/WORK/Learn/LEARN/pyrite/computer-vision/modules/eks]
ERRO[0079] 1 error occurred:
	* exit status 1
```

**Comment by bryantbiggs** [comment]
@dempti without a reproduction, I can't really help

**Comment by dempti** [comment]
> @dempti without a reproduction, I can't really help

@bryantbiggs thank you for your help and being active even in the weekends. :bow:

**Comment by AndreKR** [comment]
I'm getting this error as well. Here's my tf file:
```
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "4.15.1"
    }
  }
}

provider "aws" {
  region = "eu-central-1"
}

data "aws_availability_zones" "available" {}

module "vpc" {
  source  = "registry.terraform.io/terraform-aws-modules/vpc/aws"
  version = "3.2.0"

  name            = "my-vpc"
  cidr            = "10.0.0.0/16"
  azs             = data.aws_availability_zones.available.names
  private_subnets = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
  public_subnets  = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]

  enable_nat_gateway = true
  single_nat_gateway = true
}

module "eks" {
  source  = "registry.terraform.io/terraform-aws-modules/eks/aws"
  version = "~> 18.0"

  cluster_name = "my-cluster"

  vpc_id     = module.vpc.vpc_id
  subnet_ids = module.vpc.private_subnets

  eks_managed_node_groups = {
    main = {
      instance_types = ["t2.small"]
      capacity_type  = "SPOT"
    }
  }

  manage_aws_auth_configmap = true

  aws_auth_users = [
    {
      userarn  = "arn:aws:iam::9999999999999:user/foo"
      username = "foo"
      groups   = ["system:masters"]
    }
  ]
}
```
You need to change `userarn` and set the `AWS_...` environment variables.

**Comment by bryantbiggs** [comment]
@AndreKR  you need to add an authenticated Kubernetes provider such as:
```hcl
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id]
  }
}
```

**Comment by AndreKR** [comment]
That works, thanks.

That `exec` awscli thing felt like a massive crutch though, so I did a quick search and I found this, which seems to work as well:

```terraform
data "aws_eks_cluster_auth" "eks_auth" {
  name = module.eks.cluster_id
}

provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.eks_auth.token
}
```

**Comment by bryantbiggs** [comment]
That works as well but the `exec` method is what is recommended by the Kubernetes provider. If you exceed the token's lifetime, you will have to run a `terraform refresh`, whereas the `exec` method handles this for you

**Comment by orkramer** [comment]
Suddenly I had the exact same error. It was not before.
For me it was because the SSO token has expired.

I saw it running:
aws eks get-token --cluster-name CLUSTER

"The SSO session associated with this profile has expired."

Its good now after I refreshed the token.

**Comment by ptwohig** [comment]
I'm getting this with the module version ```~> 18.23``` and EKS version ```1.22```. Maybe I"m just missing something, but what's the solution to this.

I can create the cluster just fine, but can't modify it later because I ran into another issue were addons are stuck in the "CREATING" state.

**Comment by mconigliaro** [comment]
I think the auth map stuff belongs in a totally separate module. The [kubernetes provider documentation even warns about this](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs#stacking-with-managed-kubernetes-cluster-resources):

> When using interpolation to pass credentials to the Kubernetes provider from other resources, these resources SHOULD NOT be created in the same Terraform module where Kubernetes provider resources are also used.

**Comment by hirokistring** [comment]
FYI:
I got same error message on this issue: `Error: The configmap "aws-auth" does not exist`

But I found the cause is that I did NOT have the access to the private API endpoint of the EKS cluster.
I fixed the problem by adding `cluster_security_group_additional_rules` to allow ingress access from the place I run Terraform to the private API endpoint.

The all discussions on this issue were really helpful. Thans a lot for all.

**Comment by ctroyp** [comment]
@hirokistring would you mind posting the code segments that you used for the ingress rule (masking your ip address of course)?  I'm having the same issue.

**Comment by flomsk** [comment]
confirming its working with `token` and not with `exec` under provider configuration
```
provider "kubernetes" {
  alias                  = "eu-west-1"
  host                   = data.aws_eks_cluster.cluster_eu.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster_eu.certificate_authority.0.data)
  token                  = data.aws_eks_cluster_auth.cluster_eu.token
  # exec {
  #   api_version = "client.authentication.k8s.io/v1beta1"
  #   args        = ["eks", "get-token", "--cluster-name", data.aws_eks_cluster.cluster_eu.name]
  #   command     = "aws"
  # }
}
```

**Comment by duclm2609** [comment]
> ```terraform
> data "aws_eks_cluster_auth" "eks_auth" {
>   name = module.eks.cluster_id
> }
> ```

Me too. Using exec does not work.

**Comment by ricardo6142dh** [comment]
Still not working here :-(


<img width="726" alt="Screenshot 2022-07-03 at 23 55 59" src="https://user-images.githubusercontent.com/76172954/177058479-89355e06-3ff4-41f6-8515-231d48dbfd1e.png">


`module "eks" {
  # https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest
  source  = "terraform-aws-modules/eks/aws"
  version = "18.26.2"

  cluster_name    = local.cluster_name
  cluster_version = "1.21"

  vpc_id     = module.vpc.vpc_id
  subnet_ids = module.vpc.public_subnets

  enable_irsa = true

  create_cluster_security_group = false
  create_node_security_group    = false

  manage_aws_auth_configmap = true

  aws_auth_users = [
    {
      userarn  = "arn:aws:iam::my_aws_account:user/my_user"
      username = "lopes_becker"
      groups   = ["system:masters"]
    }
  ]`

**Comment by zeevmoney** [comment]
> in my case, I have several profiles, so I need to add `"--profile"` option.

Same thing happened to me, make sure you use the right profile.

**Comment by jeunii** [comment]
Im not sure what im doing wrong here. But my config looks like

```
provider "aws" {
  assume_role {
    role_arn = "arn:aws:iam::${var.aws_account_id}:role/pe-gitlab-assume_role"
  }
}

provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id]
  }
}

data "terraform_remote_state" "vpc" {
  backend = "http"
  config = {
    address = "https://gitlab.com/api/v4/projects/37469473/terraform/state/core-net"
   }
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"

  cluster_name    = "build"
  cluster_version = "1.22"

  cluster_endpoint_private_access = true
  cluster_endpoint_public_access  = true

  cluster_addons = {
    coredns = {}
    kube-proxy = {}
    vpc-cni = {}
  }

  vpc_id     = data.terraform_remote_state.vpc.outputs.vpc_id
  control_plane_subnet_ids = data.terraform_remote_state.vpc.outputs.dev_cp_subnet_ids

  # EKS Managed Node Group(s)
  eks_managed_node_group_defaults = {
    disk_size      = 50
    instance_types = ["t3.medium"]
    subnet_ids = data.terraform_remote_state.vpc.outputs.dev_ng_subnet_ids
  }

  eks_managed_node_groups = {
    core = {
      min_size     = 2
      max_size     = 10
      desired_size = 2

      instance_types = ["t3.2xlarge"]
      capacity_type  = "SPOT"
    }
  }

  tags = {
    ManagedBy = "Terraform"
    Infra = "eks"
  }

  # aws-auth configmap
  create_aws_auth_configmap = true

}

```

But when I run this, I get

```
│ Error: Unauthorized
│ 
│   with module.eks.kubernetes_config_map.aws_auth[0],
│   on .terraform/modules/eks/main.tf line 453, in resource "kubernetes_config_map" "aws_auth":
│  453: resource "kubernetes_config_map" "aws_auth" {
│ 
╵
Cleaning up project directory and file based variables
```

If I use just `manage_aws_auth_configmap`, I get

```
│ Error: The configmap "aws-auth" does not exist
│ 
│   with module.eks.kubernetes_config_map_v1_data.aws_auth[0],
│   on .terraform/modules/eks/main.tf line 470, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  470: resource "kubernetes_config_map_v1_data" "aws_auth" {
```

**Comment by bryantbiggs** [comment]
@jeunii you are using a role in the AWS provider which becomes the EKS default `system:master` user in the cluster, but in your kubernetes provider you are using your default profile which is probably different from the role in the IAM provider and therefore that identity does not have cluster access. You either need to update the exec call for the role assumption or use the data source route such as 

```hcl
data "aws_eks_cluster_auth" "eks_auth" {
  name = module.eks.cluster_id
}

provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.eks_auth.token
}
```

**Comment by bryantbiggs** [comment]
tl;dr - your AWS provider identity needs to be the same identity in the Kubernetes/Helm/kubectl providers when using the `exec()` method, or use the data source for the static token based auth

**Comment by jeunii** [comment]
@bryantbiggs 

thank you for your pointer. it worked. I was able to use `manage_aws_auth_configmap = true` to configure it.

when logged into the console, I see this message. 

<img width="1307" alt="Screen Shot 2022-07-19 at 5 06 37 PM" src="https://user-images.githubusercontent.com/46089687/179848556-9e2a0e1a-49b5-4f0c-a1a0-2700618d9993.png">

via the console I cannot see any nodes or k8s objects. I assume the next thing for me to do is to configure my own user ? 

```
  aws_auth_users = [
    {
      userarn  = "arn:aws:iam::66666666666:user/user1"
      username = "user1"
      groups   = ["system:masters"]
    },
]
```

I myself am a federated user via SSO.

**Comment by bryantbiggs** [comment]
correct - here's an example to add an SSO group https://github.com/clowdhaus/eks-reference-architecture/blob/main/inferentia/us-east-1/eks.tf#L30-L37

**Comment by jeunii** [comment]
@bryantbiggs Thanks. For now I just want to statically configure this.

My SSO are is `AWSReservedSSO_AdministratorAccess_05da3c667640d5fb/<firstName>.<lastName>`

so for a simple case like this, shouldnt this suffice ?

```
  # aws-auth configmap
  manage_aws_auth_configmap = true

  aws_auth_roles = [
    {
      rolearn  = "AWSReservedSSO_AdministratorAccess_05da3c667640d5fb/<firstName>.<lastName>"
      username = "<firstName>.<lastName>"
      groups   = ["system:masters"]
    },
  ]
```

**Comment by yagehu** [comment]
`exec` did not work for me while `token` with `data "aws_eks_cluster_auth` worked. Maybe in my cause it's because I'm provisioning resources from a different account.

**Comment by EnriqueHormilla** [comment]
As @bryantbiggs said it, my problem was my AWS provider is using a custom profile but my Kubernetes provider not.
```
provider "aws" {
  region  = var.region
  profile = var.aws_cli_profile  
}

provider "kubernetes" {
  host                   = module.k8s.cluster_endpoint
  cluster_ca_certificate = base64decode(module.k8s.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1beta1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.k8s.cluster_id,  "--region", var.region, "--profile", var.aws_cli_profile]
  }
}
```
After do this I can manage aws-auth with the terraform module AWESOME!!!!!

**Comment by bryantbiggs** [comment]
@jeunii you are trying to pass an entity of a role where the role ARN is required. See the example provided since this is what you are trying to accomplish. If you need to limit it further, create a custom SSO group and use that instead of the default Admin group

**Comment by ajinkyasurya** [comment]
Going to compile list of issues that I came across - 
```
╷                                                                                                                                                                                                                                        
│ Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 127.0.0.1:80: connect: connection refused                                                                                                    
│                                                                                                                                                                                                                                        
│   with kubernetes_config_map_v1_data.aws_auth[0],                                                                                                                                                                                      
│   on main.tf line 428, in resource "kubernetes_config_map_v1_data" "aws_auth":                                                                                                                                                         
│  428: resource "kubernetes_config_map_v1_data" "aws_auth" {                                                                                                                                                                            
│                                                                                                                                                                                                                                        
╵       
```
is very generic & masks multiple issues. 

1. As others have mentioned the only thing you need to get around this is following which **(didn't work)**
```
provider "kubernetes" {
  host                   = module.eks.cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)

  exec {
    api_version = "client.authentication.k8s.io/v1alpha1"
    command     = "aws"
    # This requires the awscli to be installed locally where Terraform is executed
    args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--profile", "var.profile"]
  }
}
```
I also tried with following which **(didn't work)**
```
data "aws_eks_cluster" "default" {
  name = module.eks.cluster_id
}

data "aws_eks_cluster_auth" "default" {
  name = module.eks.cluster_id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.default.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.default.token
}
```
Both the above approaches use `module.eks.cluster_id` which is not working in my case for some reason & the error masks that issue. 
**What did work was** -
```
data "aws_eks_cluster" "default" {
  name = local.cluster_name
}

data "aws_eks_cluster_auth" "default" {
  name = local.cluster_name
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.default.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.default.token
}
```

2. You'll get the same error if you have not configured `cluster_security_group_additional_rules` to make sure your cluster has correct security group to accept connections from where you are running TF. This is the second issue that's being masked. 

Try adding following to your eks module to make sure this isn't the case.

```
  cluster_security_group_additional_rules = {
    ingress = {
      description           = "To node 1025-65535"
      type                       = "ingress"
      from_port             = 0
      to_port                  = 0
      protocol                = -1
      cidr_blocks           = ["0.0.0.0/0"] # CHANGE_ME
      source_node_security_group = false
    }
  }
```

3. If you have changed cidr_blocks in 2 and you are not in the IP range you have mentioned, you might end up getting a connection timeout error -

```
╷
│ Error: Get "https://REDACT.gr7.us-west-2.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 10.X.X.X:443: i/o timeout
│ 
│   with module.eks.kubernetes_config_map_v1_data.aws_auth[0],
│   on .terraform/modules/eks/main.tf line 470, in resource "kubernetes_config_map_v1_data" "aws_auth":
│  470: resource "kubernetes_config_map_v1_data" "aws_auth" {
│ 
╵
```
Make sure you are on VPN or in the network you want to access cluster from. I'm sure there are more issues but just wanted to point out some of them that I came across.
Hopefully this saves you time!

**Comment by ctroyp** [comment]
Folks, just wanted to report back that after dealing with the "Error: The configmap "aws-auth" does not exist" error for several days, I can concur that it most likely (in my case and as experienced by others) that it was an issue in connecting to the endpoint during the process.  In my specific case, the source of my problem was with my corporate proxy where it would not accept the root certificate and terraform would report the aws-auth error and AWS CLI would report that I was using a self-signed cert which is not the case.

Hoping this will help others or at least provide some insight...

**Comment by shashidhar087** [comment]
> Going to compile list of issues that I came across -
> 
> ```
> ╷                                                                                                                                                                                                                                        
> │ Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 127.0.0.1:80: connect: connection refused                                                                                                    
> │                                                                                                                                                                                                                                        
> │   with kubernetes_config_map_v1_data.aws_auth[0],                                                                                                                                                                                      
> │   on main.tf line 428, in resource "kubernetes_config_map_v1_data" "aws_auth":                                                                                                                                                         
> │  428: resource "kubernetes_config_map_v1_data" "aws_auth" {                                                                                                                                                                            
> │                                                                                                                                                                                                                                        
> ╵       
> ```
> 
> is very generic & masks multiple issues.
> 
> 1. As others have mentioned the only thing you need to get around this is following which **didn't work**
> 
> ```
> provider "kubernetes" {
>   host                   = module.eks.cluster_endpoint
>   cluster_ca_certificate = base64decode(module.eks.cluster_certificate_authority_data)
> 
>   exec {
>     api_version = "client.authentication.k8s.io/v1alpha1"
>     command     = "aws"
>     # This requires the awscli to be installed locally where Terraform is executed
>     args = ["eks", "get-token", "--cluster-name", module.eks.cluster_id, "--profile", "var.profile"]
>   }
> }
> ```
> 
> I also tried with following which **didn't work**
> 
> ```
> data "aws_eks_cluster" "default" {
>   name = module.eks.cluster_id
> }
> 
> data "aws_eks_cluster_auth" "default" {
>   name = module.eks.cluster_id
> }
> 
> provider "kubernetes" {
>   host                   = data.aws_eks_cluster.default.endpoint
>   cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
>   token                  = data.aws_eks_cluster_auth.default.token
> }
> ```
> 
> Both the above approaches use `module.eks.cluster_id` which is not working in my case for some reason & the error masks that issue. **What did work was** -
> 
> ```
> data "aws_eks_cluster" "default" {
>   name = local.cluster_name
> }
> 
> data "aws_eks_cluster_auth" "default" {
>   name = local.cluster_name
> }
> 
> provider "kubernetes" {
>   host                   = data.aws_eks_cluster.default.endpoint
>   cluster_ca_certificate = base64decode(data.aws_eks_cluster.default.certificate_authority[0].data)
>   token                  = data.aws_eks_cluster_auth.default.token
> }
> ```
> 
> 2. You'll get the same error if you have not configured `cluster_security_group_additional_rules` so make sure your cluster has correct security group to accept connections from where you are running terraform. This is the second issue that's being masked.
> 
> Try adding following to your eks module to make sure this isn't the case.
> 
> ```
>   cluster_security_group_additional_rules = {
>     ingress = {
>       description           = "To node 1025-65535"
>       type                       = "ingress"
>       from_port             = 0
>       to_port                  = 0
>       protocol                = -1
>       cidr_blocks           = ["0.0.0.0/0"] # CHANGE_ME
>       source_node_security_group = false
>     }
>   }
> ```
> 
> 3. If you have changed cidr_blocks in 2 and you are not in the IP range you have mentioned, you might end up getting a connection timeout error -
> 
> ```
> ╷
> │ Error: Get "https://REDACT.gr7.us-west-2.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 10.X.X.X:443: i/o timeout
> │ 
> │   with module.eks.kubernetes_config_map_v1_data.aws_auth[0],
> │   on .terraform/modules/eks/main.tf line 470, in resource "kubernetes_config_map_v1_data" "aws_auth":
> │  470: resource "kubernetes_config_map_v1_data" "aws_auth" {
> │ 
> ╵
> ```
> 
> Make sure you are on VPN or in the network you want to access cluster from. I'm sure there are more issues but just wanted to point out some of them that I came across. Hopefully this saves you time!

My issue solved. Thanks !

**Comment by shake76** [comment]
Thanks all for sharing this information and workarounds , actually I'm having the following issue "https://68E51D9223F166C1D6B7BDCDFD370F1F.gr7.region.eks.amazonaws.com/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp 10.x.x.x:443: i/o timeout", it happens when run Terraform Cloud plan against eks module after turn on private endpoints in my EKS cluster, I already configured a bastion host that is able to reach our cluster, from my local I've configured a proxy that use ssm, even more  I was able to run kubectl commands without problem to interact with the cluster but the issue is coming when run terraform cloud plan, Do you know if is there any workaround to pass an https_proxy variable with Terraform Cloud? I'll appreciate your comments in advance

**Comment by yagehu** [comment]
Another cause: I had set `endpoint_public_access  = false`. In that case I would need an SSH tunnel.

**Comment by VladoPortos** [comment]
This is a constant headache:
```
│ Error: configmaps "aws-auth" already exists
│ 
│   with module.eks.kubernetes_config_map.aws_auth[0],
│   on .terraform/modules/eks/main.tf line 453, in resource "kubernetes_config_map" "aws_auth":
│  453: resource "kubernetes_config_map" "aws_auth" {
```

The map is there because I have:

```
  manage_aws_auth_configmap = true
  create_aws_auth_configmap = true
```

But it looks like it want to create the aws map on every run. I did not have yet successful run in creating EKS cluster with Terraform. This issue always pops up. Only solution is to import the map, and then re run again... not very automation friendly if I have to do manual task :(

**Comment by yagehu** [comment]
> This is a constant headache:
> 
> ```
> │ Error: configmaps "aws-auth" already exists
> │ 
> │   with module.eks.kubernetes_config_map.aws_auth[0],
> │   on .terraform/modules/eks/main.tf line 453, in resource "kubernetes_config_map" "aws_auth":
> │  453: resource "kubernetes_config_map" "aws_auth" {
> ```
> 
> The map is there because I have:
> 
> ```
>   manage_aws_auth_configmap = true
>   create_aws_auth_configmap = true
> ```
> 
> But it looks like it want to create the aws map on every run. I did not have yet successful run in creating EKS cluster with Terraform. This issue always pops up. Only solution is to import the map, and then re run again... not very automation friendly if I have to do manual task :(

I think you just need to set `create_aws_auth_configmap` to false. This is because `aws-auth` config map is automatically created by EKS.

**Comment by yagehu** [comment]
I've encountered and fixed this issue multiple times. This is mainly because we provision with Terraform from a different account. The crux of the issue is:

1. On creation, EKS only allows the *IAM entity* that created the cluster access.
2. If provisioning cross-account, the entity that created the cluster is the *role* that is assumed.
3. You have to use that role to change the `aws-auth` config map or perform

To know definitively what entity it is, check your CloudTrail events and search for event name `CreateCluster`, for example:

```
{
    // snip
    "userIdentity": {
        "type": "AssumedRole",
        // snip
        "sessionContext": {
            "sessionIssuer": {
                "type": "Role",
                // snip
                "arn": "arn:aws:iam::0123456789:role/xxxxxx",  <------- This is the role
                // snip
            },
        }
    }
}
```

**Comment by dantman** [comment]
For users running into this issue and having issues with the exec provider. Also try using `aws eks update-kubeconfig ...` and connecting to your cluster directly. Or just check your aws cli version.

You could be running into issues like https://github.com/aws/aws-cli/issues/6920 which can be caused, among other things, by having v1 of the aws cli installed from an apt package instead of a recent v2 version.

The kubernetes provider doesn't seem to let you know when it couldn't connect because your aws cli is out of date.

**Comment by VladoPortos** [comment]
I needed to add this to get it to work, finally, 

```
resource "null_resource" "kubectl-init" {
  provisioner "local-exec" {
    command = "aws eks --region ${var.aws_region} update-kubeconfig --name ${var.cluster_name}"
  }
  depends_on = [module.eks.cluster_id]
}
```
This needs to run after eks creation, before module installation.
This will populate the cubeconfig needed by exec later on. After this, it worked ok

**Comment by kingbj940429** [comment]
### If you've tried all of the above methods, but you can't, try the ones I wrote.

I have same problem

`Error: The configmap "aws-auth" does not exist`

But I SOVLED this way !! 

First, I had to create EKS Cluster API Endpoint [Private]

As it says in AWS Docs, you can only access EKS Cluster API-Server in Same VPC.

My Company is using TGW. so I Can Access VPC where created EKS Cluster.

Anyway, the reason that occur  a error is you couldn't access EKS Api-server !!

# Solve

#### I proceed thinking that you are in the same VPC as EKS Cluster
<img width="1395" alt="스크린샷 2022-09-29 오후 7 23 44" src="https://user-images.githubusercontent.com/63000843/193009541-30ac991c-ed15-41a9-bba9-cacf3ab02b14.png">

click additional security group !!

<img width="1425" alt="스크린샷 2022-09-29 오후 7 25 20" src="https://user-images.githubusercontent.com/63000843/193009664-e9fdff28-ebdc-492b-9365-090b71a7090c.png">

and add rule like image !! allow 443 port
``` 
 # Extend cluster security group rules
  cluster_security_group_additional_rules = {
    ingress = {
      description                = "EKS Cluster allows 443 port to get API call"
      type                       = "ingress"
      from_port                  = 443
      to_port                    = 443
      protocol                   = "TCP"
      cidr_blocks                = ["0.0.0.0/0"]
      source_node_security_group = false
    }
  }
```
Even if it's private, you can access eks cluster Api-server and can get aws-auth in kube-system !!

By opening the 443 port, it allows you to connect to the EKS Api-server.

Based on the EKS Api-server, 443 ports were closed, so you couldn't access it, and this error occurred.

**Comment by dejongm** [comment]
We resolved our issue with a `depends_on` in the `aws_eks_cluster` and `aws_eks_cluster_auth` data blocks. I believe the K8s creds are expiring while the node groups are being created. We don't use the terraform-aws-eks module but our data blocks looks like this:
```
module "eks_node_group" {
...
}

data "aws_eks_cluster" "main" {
  depends_on = [
    module.eks_node_group
  ]
  name = aws_eks_cluster.main.id
}

data "aws_eks_cluster_auth" "main" {
  depends_on = [
    module.eks_node_group
  ]
  name = aws_eks_cluster.main.id
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.main.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.main.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.main.token
}


provider "helm" {
  kubernetes {
    host                   = data.aws_eks_cluster.main.endpoint
    cluster_ca_certificate = base64decode(aws_eks_cluster.main.certificate_authority[0].data)
    token                  = data.aws_eks_cluster_auth.main.token
  }
}
```

**Comment by tulanian** [comment]
Similar to this [comment](https://github.com/terraform-aws-modules/terraform-aws-eks/issues/2009#issuecomment-1272089141)  we solved this by adding depndencies. The data sources aren't populated until after all of the node groups have been created, so the token is fresher.

`data "aws_eks_cluster" "this" {
  name = module.eks.cluster_id
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

data "aws_eks_cluster_auth" "this" {
  name = module.eks.cluster_id
  depends_on = [
    module.eks.eks_managed_node_groups,
  ]
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.this.endpoint
  token                  = data.aws_eks_cluster_auth.this.token
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.this.certificate_authority.0.data)
}

`

**Comment by gartemiev** [comment]
I've face with the same issue. However - I had outdated aws cli version and aws-iam-authenticator. 
Once I updated aws cli and aws-iam-authenticator along with kube confing  - the issue has resolved.
